torchdynamo_poc/TorchDynamo Demo.ipynb

```diff
@@ -309,29 +378,79 @@
-      "    %0 = torch.aten.add.Scalar %arg0, %int1, %int1 : !torch.vtensor<[2,3],f32>, !torch.int, !torch.int -> !torch.vtensor<[2,3],f32>\n",
-      "    %1 = torch.aten.div.Tensor %arg0, %0 : !torch.vtensor<[2,3],f32>, !torch.vtensor<[2,3],f32> -> !torch.vtensor<[2,3],f32>\n",
-      "    %2 = torch.aten.sum %arg1, %none : !torch.vtensor<[2,3],f32>, !torch.none -> !torch.vtensor<[],f32>\n",
-      "    %3 = torch.aten.lt.Scalar %2, %int0 : !torch.vtensor<[],f32>, !torch.int -> !torch.vtensor<[],i1>\n",
-      "    return %1, %3 : !torch.vtensor<[2,3],f32>, !torch.vtensor<[],i1>\n",
+      "    %int2 = torch.constant.int 2\n",
+      "    %0 = torch.aten.div.Scalar %arg0, %int2 : !torch.vtensor<[2,3],f32>, !torch.int -> !torch.vtensor<[2,3],f32>\n",
+      "    %1 = torch.aten.max %arg1 : !torch.vtensor<[2,3],f32> -> !torch.vtensor<[],f32>\n",
+      "    %2 = torch.aten.lt.Scalar %1, %int0 : !torch.vtensor<[],f32>, !torch.int -> !torch.vtensor<[],i1>\n",
+      "    return %0, %2 : !torch.vtensor<[2,3],f32>, !torch.vtensor<[],i1>\n",
       "  }\n",
       "}\n",
       "\n",
+      "Compiling graph...\n",
+      "torch.fx graph:\n",
+      "graph():\n",
+      "    %arg0_1 : [#users=1] = placeholder[target=arg0_1]\n",
+      "    %arg1_1 : [#users=1] = placeholder[target=arg1_1]\n",
+      "    %mul : [#users=1] = call_function[target=torch.ops.aten.mul](args = (%arg1_1, %arg0_1), kwargs = {})\n",
+      "    return mul\n",
+      "\n",
+      "\n",
+      "torch-mlir backend contract graph:\n",
+      "module attributes {torch.debug_module_name = \"_lambda\"} {\n",
+      "  func.func @forward(%arg0: !torch.vtensor<[2,3],f32>, %arg1: !torch.vtensor<[2,3],f32>) -> !torch.vtensor<[2,3],f32> {\n",
+      "    %0 = torch.aten.mul.Tensor %arg1, %arg0 : !torch.vtensor<[2,3],f32>, !torch.vtensor<[2,3],f32> -> !torch.vtensor<[2,3],f32>\n",
+      "    return %0 : !torch.vtensor<[2,3],f32>\n",
+      "  }\n",
+      "}\n",
+      "\n"
+     ]
+    },
+    {
+     "data": {
+      "text/plain": [
+       "tensor([[0.5000, 0.5000, 0.5000],\n",
+       "        [0.5000, 0.5000, 0.5000]])"
+      ]
+     },
+     "execution_count": 13,
+     "metadata": {},
+     "output_type": "execute_result"
+    }
+   ],
+   "source": [
+    "foo(torch.ones((2, 3)), torch.ones((2, 3)))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "349c7526",
+   "metadata": {},
+   "source": [
+    "Notice that because the body of the `if` statement is not used, it is not compiled. If we change the inputs so that the body of the `if` statement is needed, then TorchDynamo will reuse the compiled graphs for the other parts of the function and only compile the body of the `if` statement."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": 14,
+   "id": "3280d186",
+   "metadata": {},
+   "outputs": [
+    {
+     "name": "stdout",
+     "output_type": "stream",
+     "text": [
       "Compiling graph...\n",
       "torch.fx graph:\n",
       "graph():\n",
```

Comment:
These graphs are quite hard to follow, but I don't have any good ideas on how to improve them here.  Seems like it might need a dedicated deep dive in a separate doc (though maybe that's too low level for anyone but us?).  In particular I can't tell how the sub-graphs / functions are connected.

Reply:
Yeah, there is some assumed familiarity with torch.fx graphs and MLIR graphs. The graph right under this comment

```
graph():
    %arg0_1 : [#users=1] = placeholder[target=arg0_1]
    %arg1_1 : [#users=1] = placeholder[target=arg1_1]
    %mul : [#users=1] = call_function[target=torch.ops.aten.mul](args = (%arg0_1, -1), kwargs = {})
    %mul_1 : [#users=1] = call_function[target=torch.ops.aten.mul](args = (%arg1_1, %mul), kwargs = {})
    return mul_1
```

represents the body of the `if` statement plus the return statement:

```python
b = b * -1
return x * b
```

since this path is now needed given the inputs.

From the previous cell, the graph

```
graph():
    %arg0_1 : [#users=1] = placeholder[target=arg0_1]
    %arg1_1 : [#users=1] = placeholder[target=arg1_1]
    %div : [#users=1] = call_function[target=torch.ops.aten.div](args = (%arg0_1, 2), kwargs = {})
    %max_1 : [#users=1] = call_function[target=torch.ops.aten.max](args = (%arg1_1,), kwargs = {})
    %lt : [#users=1] = call_function[target=torch.ops.aten.lt](args = (%max_1, 0), kwargs = {})
    return (div, lt)
```

represents the part of the cumputation up until the `if` statement.

```python
x = a / 2
lt = b.max() < 0
return x, lt
```

and the second graph in that output represents the return statement:

```
graph():
    %arg0_1 : [#users=1] = placeholder[target=arg0_1]
    %arg1_1 : [#users=1] = placeholder[target=arg1_1]
    %mul : [#users=1] = call_function[target=torch.ops.aten.mul](args = (%arg1_1, %arg0_1), kwargs = {})
    return mul
```

```python
return x * b
```

Reply:
That makes sense.  I think the piece I'm missing is how these graphs are stitched together.  Is there another graph that's not shown that takes the results of the if statement graph and decides which other graph to call, for example?